### ISSUE TITLE
False positive on unnecessary_unwrap where more than one value gets checked then unwrapped

### ISSUE THREAD

**Issue by rscarson** [rust-lang/rust-clippy#11513]
### Summary

As the title states, #unnecessary_unwrap seems to result in a false positive in cases where 2 or more values get checked / unwrapped

Though this:
```rust
if a.is_none() || b.is_none() {
    handle_error()
} else {
    handle(a.unwrap(), b.unwrap())
}
```

Could be written to avoid this lint as:
```rust
if let Some(va) = a {
    if let Some(vb) = b {
        handle(va, vb)
    } else {
        handle_error()
    }
} else {
    handle_error()
}
```

That version is far more verbose, and results in duplicating error handling.

I think if the same block triggers this lint more than once, it should be ignored all together

### Lint Name

unnecessary_unwrap

### Reproducer

I tried this code:

```rust
if left.is_none() || left.is_none() {
    // Handle the error
} else {
    // Call handler
    handler(left.unwrap(), left.unwrap())
}
```

I saw this happen:

```
warning: called `unwrap` on `left` after checking its variant with `is_none`
  --> src\handlers\utils.rs:65:27
   |
61 |         if left.is_none() || right.is_none() {
   |                            ------------ the check is happening here
...
65 |             match handler(left.unwrap(), right.unwrap()) {
   |                                        ^^^^^^^^
   |
   = help: try using `if let` or `match`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap

warning: called `unwrap` on `right` after checking its variant with `is_none`
  --> src\handlers\utils.rs:65:40
   |
61 |         if left.is_none() || right.is_none() {
   |                            ------------ the check is happening here
...
65 |             match handler(left.unwrap(), right.unwrap()) {
   |                                                            ^^^^^^^^^
   |
   = help: try using `if let` or `match`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap
```

I expected to see this happen:

both `if let` and `match` would result in a larger, more redundant implementation - if there are 2 or more conditions being checked, this one shouldn't pop

### Version

```text
$ rustc -Vv
rustc 1.72.0 (5680fa18f 2023-08-23)
binary: rustc
commit-hash: 5680fa18feaa87f3ff04063800aec256c3d4b4be
commit-date: 2023-08-23
host: x86_64-pc-windows-msvc
release: 1.72.0
LLVM version: 16.0.5

$ cargo clippy --version
clippy 0.1.72 (5680fa18 2023-08-23)
```

**Comment by Jarcho** [comment]
You can also use a tuple here. e.g. `if let (Some(_), Some(_)) = (x, y) { ... }`

**Comment by rscarson** [comment]
Oh, well TIL

Maybe the error could suggest that option in cases like this? I didn't even know that syntax existed

**Comment by y21** [comment]
Note that it can change the behavior if `y` is an expression with side effects and not just a reference to a local, since it now always evaluates it and no longer short circuits if `x` is `None`, but maybe we can use clippy's [lazy to eager util](https://doc.rust-lang.org/stable/nightly-rustc/clippy_utils/eager_or_lazy/fn.switch_to_eager_eval.html) to see if eager evaluation wouldn't make a difference and is applicable.

There's also `if let Some((x, Some(y)) = x.map(|x| (x, y)) {}`, which maintains the short circuit but is specific to `Option<_>` and doesn't look as nice

**Comment by y21** [comment]
Ah, my comment might not be really relevant, because it looks like the lint already only fires on local path expressions anyway..

**Comment by rscarson** [comment]
I'd still love to see an update to the docs suggesting the `if let (Some(_), Some(_)) = (x, y) { ... }` syntax in cases like these where skipping short-circuit eval would be unlikely to have side effects

**Comment by rscarson** [comment]
Actually just ran into another example of this issue in the wild:
```rust
if value.is_none() || value.unwrap().is_undefined() {
    Err(ValueNotFoundError::new(name.to_string()).into())
} else {
    Ok(v8::Global::<v8::Value>::new(&mut scope, value.unwrap()))
}
```

A combination of a a check to is_none and a second non-let condition also triggered it

**Comment by rscarson** [comment]
Any update on this?

**Comment by J-ZhengLi** [comment]
So... Just providing a more helpful suggestion is enought right? The only concern should be the suggestion won't be auto-fixed as a new variable name is needed. But I think It should be alright since there are some lints offering suggestion with (possibly invalid) variable names as well, such as [clippy::blocks_in_conditions](https://rust-lang.github.io/rust-clippy/master/index.html#/blocks_in_conditions) @rustbot claim
